ScaLAPACK-like interface for Cholesky decomposition and eigensolver

RMeli opened the pull request:
Introduce a ScaLAPACK-like interface to simplify adoption of DLA-Future, as a drop-in replacement.

This PR is based on work done to integrate DLA-Future in CP2K.

* [x] Introduce ScaLAPACK as optional dependency with ~`-DDLAF_WITH_INTERFACE`~ ~`-DDLAF_WITH_C_API`~ `-DDLAF_WITH_SCALAPACK`
  * [x] MKL
  * [x] Other
* [x] Introduce ~`+interface`~ ~`+c`~ `+scalapack` variant in Spack package
  * [x] MKL
  * [x] Other
* [x] Functions for `pika` and `dlaf` initialization/finalization  
* [x] Functions for extracting `Distribution`, `LayoutInfo`, and `CommunicationGrid` from ScaLAPACK descriptors (and BLACS contexts)
* [x] Some BLACS functions
* [x] Interface for Cholesky decomposition
  * [x] `float`
  * [x] `double`
  * [x] `complex float`
  * [x] `complex double`
* [x] Interface for eigensolver
  * [x] `float`
  * [x] `double`
  * [x] `complex float`
  * [x] `complex double`
* [x] Use `MPI_Thread_Init` in MPI (non-pika) tests
* [x] Tests
  * [x] Cholesky
  * [x] Eigensolver
* [x] Documentation

~This PR is very preliminary, there is still a lot of work and cleanup to do. This is especially true for the one test of the Cholesky decomposition interface, which are very messy and "hand-made" right now (quick-and-dirty solution to re-create something similar to what we have in CP2K).~

I'm opening it as a draft PR early to facilitate discussion (about naming conventions, position within the code base, better design patterns, ...).

Early feedback is very welcome as usual.

### Notes

* ~[ ] @rasolca suggests to implement something similar to `MatrixLocal AllGather(...)` to re-use already implemented checks~
  * *Re-used DLA-Future functionality for creating the matrix and performing the tests (requires careful use of `pika::resume` and `pika::suspend`)*  

---

A lot of credit goes to @mtaillefumier and @msimberg, who worked on the initial integration of the Cholesky decomposition in CP2K.

---

Spack env:
```yaml
spack:
  specs:
  - dla-future@develop +miniapps +cuda cuda_arch=89 ^intel-mkl threads=openmp ^mpich
    device=ch3 netmod=tcp ^pika@0.15.1 ^umpire~device_alloc
  view: false
  packages:
    all:
      variants:
      - cxxstd=17
  repos:
  - ~/git/DLA-Future/spack
  concretizer:
    unify: true
```

Build env:
```bash
spack -e ~/git/my-spack/envs/local/dlaf/mkl-mt-mpich-cuda build-env dla-future -- bash
```

CMake:
```bash
cmake -DDLAF_WITH_CUDA=on -DDLAF_WITH_INTERFACE=on -DDLAF_WITH_MKL=on -DMKL_LAPACK_TARGET=mkl::mkl_intel_32bit_omp_dyn -DMKL_SCALAPACK_TARGET=mkl::scalapack_mpich_intel_32bit_omp_dyn -DDLAF_MPI_PRESET=plain-mpi -DMPIEXEC_NUMCORES=24 -DCMAKE_EXPORT_COMPILE_COMMANDS=1   ..
```

---

Fix #610

RMeli commented on src/CMakeLists.txt:
```diff
@@ -75,6 +75,7 @@ target_link_libraries(
     MPI::MPI_CXX
     $<$<BOOL:${DLAF_WITH_OPENMP}>:OpenMP::OpenMP_CXX>
     DLAF::LAPACK
+    DLAF::SCALAPACK
```
```suggestion
    $<$<BOOL:${DLAF_WITH_INTERFACE}>:DLAF::SCALAPACK>
```

RMeli commented on test/unit/interface/test_cholesky.cpp:
```diff
@@ -0,0 +1,165 @@
+
+#include <mpi.h>
+
+#include <iostream>
```
```suggestion
```
Eventually...

rasolca commented:
Just a few thoughts to start a discussion:
1. I would use `dlaf_C` instead of `interface`
2. Do we want to separate the C interface? Another option can be to add it directly in the tree.
e.g.
Currently
- factorization.h
- factorization 
  - cholesky.h
with C wrappers
- factorization.h
- factorization_C.h
- factorization 
  - cholesky.h
  - cholesky_C.h

3. We should avoid creating and destroying the Comm grid at each DLAF call. We should think of a way to reuse it.

4. Some considerations about the descriptor:
  - In SCALAPACK the descriptor describes the full matrix, however `ia`, `ja` and `n` (I took Cholesky as example) describes the submatrix used in the algorithm. We should introduce a DLAF descriptor to describe the actual matrix used in the algorithm. It can be a simple struct with the following entries:
```
struct DLAF_descriptor {
Int m;
Int n;
Int mb;
Int nb;
Int isrc;
Int jsrc;
Int i; // currently not supported by DLAF (has to be 0, adding it so we can assert.) 
Int j; // same as i
Int ld;
};
```
(Note: which `Int` type?)

msimberg commented on src/interface/cholesky.cpp:
```diff
@@ -0,0 +1,31 @@
+
+namespace dlaf::interface {
+
+extern "C" void pdpotrf(char uplo, int n, double* a, int ia, int ja, int* desca, int& info) {
```
It's probably enough to put `extern "C"` on the definition, right? Assuming these are all lifted out of the `dlaf::interface` namespace.

RMeli replied:
Yes, I think so.

msimberg commented on include/dlaf/interface/cholesky.h:
```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <mpi.h>
+#include <dlaf/interface/cholesky.h>
```
Including self?

msimberg commented on include/dlaf/interface/cholesky.h:
```diff
@@ -0,0 +1,105 @@
+//
```
I would definitely separate out the actual C interface functions into a separate header so that they can actually be included in C files without dragging in all of dlaf and pika (which won't work anyway).

msimberg reviewed (changes requested):
Thanks @RMeli for taking this on, this looks already really good.

The main comment I have (as @rasolca) is about the naming. The C interface should be in the top-level namespace and I would name the functions `dlaf_*` (`dlaf_C_*` feels redundant since we're anyway talking about the C interface).

Regarding organization, a couple of other options are:
- wrap all C++ functionality in `#ifdef __cplusplus` and leave the C interface outside the ifdef; this way we could have the C interface in the same header files
- rename the C++ header files to use `.hpp` as the extension and provide a separate, but corresponding set of headers with `.h` extensions for the C interface

RMeli commented on include/dlaf/interface/cholesky.h:
```diff
@@ -0,0 +1,105 @@
+  info = 0;
+}
+
+extern "C" void cholesky_d(char uplo, double* a, int m, int n, int mb, int nb, int lld,
+                           const MPI_Comm& communicator, int nprow, int npcol);
+
+extern "C" void cholesky_s(char uplo, float* a, int m, int n, int mb, int nb, int lld,
+                           const MPI_Comm& communicator, int nprow, int npcol);
+
+extern "C" void pdpotrf(char uplo, int n, double* a, int ia, int ja, int* desca, int& info);
+
+extern "C" void pspotrf(char uplo, int n, float* a, int ia, int ja, int* desca, int& info);
```
Sounds good for Cholesky, but what about `dlaf_eigensolver_eigensolver_d`? I have a slight preference for the former: `dlaf_cholesky_d`, `dlaf_eigensolver_d`, ...

msimberg replied:
> but what about `dlaf_eigensolver_eigensolver_d`

Yeah, that doesn't look great... Is it a bad idea to come up with some naming scheme close to `pdpotrf` for the first two variants or is that going to cause confusion?

RMeli replied:
I think it might be worth differentiating the ScaLAPACK-like API (which needs ScaLAPACK as a dependency, so we might want to make it conditional)? For now I went with the following, but open to further discussion:
```
dlaf_pdpotrf(uplo, n, a, ia, ja, desc_a, info)
dlaf_cholesky_d(dlaf_context, uplo, a, dlaf_desc_a)
```

rasolca commented on include/dlaf/interface/blacs.h:
```diff
@@ -0,0 +1,33 @@
+#include <mpi.h>
+
+namespace dlaf::interface::blacs {
+// Cblacs does not have public header files and is only used in ScaLAPACK
```
The interface of blacs is public, and `Cblacs.h` should be in general available. Am I wrong?

RMeli replied:
This is a leftover comment from @mtaillefumier, which could provide some additional context?

rasolca commented on include/dlaf/interface/cholesky.h:
```diff
@@ -0,0 +1,105 @@
+template <typename T>
+using MatrixMirror = dlaf::matrix::MatrixMirror<T, dlaf::Device::Default, dlaf::Device::CPU>;
+
+template <typename T>
+void pxpotrf(char uplo, T* a, int m, int n, int mb, int nb, int lld, const MPI_Comm& communicator,
+             int nprow, int npcol) {
```
This should be an internal algorithm API, that is used by the C interface.
Anyway it is C++, so it can be declared and defined in `include/dlaf/cholesky.h`

RMeli replied:
Moved to private headers in `src/c_api`, as suggested below.

> Any private interface (e.g. the templated pxpotrf) should be added either to the source files or to private headers in the src/c_api tree.

rasolca commented on include/dlaf/interface/cholesky.h:
```diff
@@ -0,0 +1,105 @@
+template <typename T>
+void pxpotrf(char uplo, T* a, int m, int n, int mb, int nb, int lld, const MPI_Comm& communicator,
+             int nprow, int npcol) {
+  pika::resume();
```
I would not support calling multiple algorithm with the C interface. It anyway needs a synchronization as the matrix object should be recreated at each call.

msimberg replied:
Right, makes sense. In any case, I would simply document that we initialize and start the pika runtime in `dlaf_initialize`, resume and suspend the runtime around algorithm calls, and stop the pika runtime in `dlaf_finalize`, since this affects composability with other libraries potentially using pika. I know it's not a concern right now (or ever!) but I'd just like it to be explicitly documented.

RMeli commented on .gitignore:
```diff
@@ -87,3 +87,6 @@ venv/
 ENV/
 env.bak/
 venv.bak/
+
+# ------------------- Spack
+/spack-*
```
https://github.com/eth-cscs/DLA-Future/pull/891

rasolca commented on test/unit/interface/test_cholesky.cpp:
```diff
@@ -0,0 +1,303 @@
+//
```
Not sure... if we want to use gtest I think we need c++.

RMeli replied:
I think here using `.c` would be more useful than using GTest, since we want to test that everything work as expected when using the interface, while correctness tests are already handled extensively elsewhere.

msimberg replied:
Yeah, obviously things like GTest may need a C++ compiler. However, I'd say even just making a dummy `.c` file which does something silly like:
```
void my_cholesky_d(char uplo, double* a, int m, int n, int mb, int nb, int lld,
                           const MPI_Comm& communicator, int nprow, int npcol) {
    cholesky_d(uplo, a, m, n, mb, nb, lld, communicator, nprow, npcol);
}
```
and then calling `my_cholesky_d` from a C++ file would be better in my opinion just to make sure the C interface can be used in a C file. Note, there may be a better way of doing this, the above is just an example.

rasolca replied:
`Everything works` doesn't mean `everything works correctly`.
A small typo in one when setting one of the sizes in the interface is enough to break everything silently.

RMeli replied:
Yes, I was not suggesting not to test for correctness, but to have a smaller set of C tests with less automation than provided by GTest. Something in the lines of what @msimberg suggests seems to combine the best of both worlds, I'll look into it.

RMeli replied:
I think this is fixed now, I followed @msimberg suggestion of having a dummy `.c` file and then used GTest for the actual tests (with most of the code shared with the standard tests).

RMeli commented on CMakeLists.txt:
```diff
@@ -43,6 +43,7 @@ option(DLAF_BUILD_MINIAPPS "Build miniapps" ON)
 option(DLAF_BUILD_TESTING "Build tests" ON)
 option(DLAF_BUILD_DOC "Build documentation" OFF)
 option(DLAF_WITH_PRECOMPILED_HEADERS "Use precompiled headers." OFF)
+option(DLAF_WITH_INTERFACE "Build ScaLAPACK-like interface" OFF)
```
Good point, I used "interface" for lack of better ideas and inspired by DLA-Interface. Which suggestion people prefer? (I don't mind, but I'd vote slightly in favor of `DLAF_WIT_C_API`.)

msimberg replied:
Yep, completely reasonable! I'm growing on `DLAF_WITH_C_API` as well, but not sure yet. Curious to hear what the others think.

msimberg commented on include/dlaf/interface/eigensolver.h:
```diff
@@ -0,0 +1,77 @@
+namespace dlaf::interface {
+
+template <typename T>
+using MatrixMirror = dlaf::matrix::MatrixMirror<T, dlaf::Device::Default, dlaf::Device::CPU>;
```
Oh right, we use that in DLA-Future... At least we've commented that HIP doesn't have something similar (https://github.com/eth-cscs/DLA-Future/blob/9fcd6d0bee0817b06557fc9cbc8532966d301994/include/dlaf/eigensolver/band_to_tridiag/mc.h#L395-L398).

I'm of course asking in case we can either:

1. Check that the allocated memory is on the expected device, or
2. Copy the matrix to a GPU if needed.

However, these can be extras. For now we should at least document that we expect the input matrix to be in CPU memory.

RMeli replied:
This has been documented. Do we need to make a note somewhere to re-check this if HIP support becomes available?

RMeli commented:
Thanks @rasolca and @msimberg for all the suggestions and the early review. I implemented most of them in case you want to have a second look while I focus on the tests. I went with `int` as `DLAF_ContextType` for now, but it can be changed later if needed. 

I removed the eigensolver for now, to avoid having to modify things twice; I'll add the eigensolver when we are happy with Cholesky.

msimberg commented on include/dlaf_c/desc.h:
```diff
@@ -0,0 +1,22 @@
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
```
```suggestion
//

```

msimberg commented on include/dlaf_c/factorization/cholesky.h:
```diff
@@ -0,0 +1,30 @@
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
```
```suggestion
//

```

msimberg commented on include/dlaf_c/grid.h:
```diff
@@ -0,0 +1,22 @@
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
```
```suggestion
//

```

msimberg commented on include/dlaf_c/grid.h:
```diff
@@ -0,0 +1,22 @@
+#include <mpi.h>
+
+extern "C" {
+MPI_Comm Cblacs2sys_handle(int ictxt);
+void Cblacs_get(int ictxt, int inum, int* comm);
+extern "C" void Cblacs_gridinfo(int ictxt, int* np, int* mp, int* px, int* py);
```
What was the conclusion about these? Can they not be accessed through a scalapack header? It looks like it should at least be possible to move them to `grid.cpp` if they need to be declared?

mtaillefumier replied:
The official scalapack does not have header files (please correct me if I am wrong). I searched in the code and the BLACS library which is written in C does not export functions and structures. scalapack been in FORTRAN, only export modules files (when we are lucky) only. 
@RMeli  pointed out in a private message that MKL provides an interface which is good but all the other implementations do not necessarily. the Cblacs functions should be declared explicitly in that case. COSMA does the same (i took it from cosma initially).

RMeli replied:
See comment below.

msimberg commented on include/dlaf_c/init.h:
```diff
@@ -0,0 +1,15 @@
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
```
```suggestion
//

```

msimberg commented on include/dlaf_c/init.h:
```diff
@@ -0,0 +1,15 @@
+//
+#pragma once
+
+extern "C" {
```
Does this need `#ifdef __cplusplus`?

msimberg commented on src/c_api/grid.h:
```diff
@@ -0,0 +1,15 @@
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
```
```suggestion

#pragma once
```

msimberg commented on test/unit/factorization/test_cholesky_c_api_wrapper.h:
```diff
@@ -0,0 +1,12 @@
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
```
```suggestion

#pragma once
```

msimberg commented on test/unit/factorization/test_cholesky_c_api.cpp:
```diff
@@ -0,0 +1,316 @@
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
```
```suggestion
//

```

msimberg commented on test/unit/factorization/test_cholesky_c_api.cpp:
```diff
@@ -0,0 +1,316 @@
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+extern "C" {
```
Move this into the header file?

msimberg commented on src/c_api/init.cpp:
```diff
@@ -0,0 +1,47 @@
+
+static bool dlaf_initialized = false;
+
+void dlaf_initialize(int argc, const char** argv) {
```
```suggestion
extern "C" {
void dlaf_initialize(int argc, const char** argv) {
```

or my preference would be to only have the `extern "C"` on the declaration. Whichever we do, we should be consistent. Currently some of the C API _definitions_ use `extern "C"` (`dlaf_pdpotrf` etc.) and some don't (`dlaf_initialize` etc.).

Semi-related, it may be worth defining a macro:
```
#ifdef __cplusplus
#define DLAF_EXTERN_C
#else
#define DLAF_EXTERN_C extern "C"
#endif
```
and use that on the declarations instead.

RMeli replied:
The macro is a good suggestion, thanks.

msimberg reviewed (changes requested):
This is looking a lot better, thanks @RMeli for the updates! Remaining comments are mostly cosmetic. There are some `TODO`s that we shouldn't forget to deal with either.

RMeli commented:
> There are some TODOs that we shouldn't forget to deal with either.

Yes, I should have mentioned that I just wanted to check that the refactoring based on the suggestions is going on the right direction. Still a lot of TODOs... =)

msimberg commented on include/dlaf_c/grid.h:
```diff
@@ -0,0 +1,24 @@
+
+#include "dlaf_c/utils.h"
+
+DLAF_EXTERN_C MPI_Comm Cblacs2sys_handle(int ictxt);
+DLAF_EXTERN_C void Cblacs_get(int ictxt, int what, int* val);
+DLAF_EXTERN_C void Cblacs_gridinfo(int ictxt, int* np, int* mp, int* px, int* py);
```
These can maybe be put in a separate header included only for the internal implementations? We wouldn't want users to rely on us declaring these for them.

RMeli replied:
Good suggestion. I'm still debating if we should include `mkl_blacs.h` for MKL and define them only for other implementations. Or just do what COSMA seems to do and define them regardless.

RMeli replied:
> These can maybe be put in a separate header included only for the internal implementations?

Done. But I'll leave this unresolved until I have checked if for MKL we cam/should just use `mkl_blacs.h`.

msimberg commented on spack/packages/dla-future/package.py:
```diff
@@ -25,11 +25,15 @@ class DlaFuture(CMakePackage, CudaPackage, ROCmPackage):
 
     variant("miniapps", default=False, description="Build miniapps.")
 
+    variant("c_api", default=False, description="Build C API")
```
I don't feel strongly about the name of this variant, but I had a quick look in spack and found:
- `variant("c", ...)`: 8 occurrences
- `variant("c_interface", ...)`: 1 occurrence
- `variant("c_shared", ...)`: 1 occurrence (not quite the same)

There were also 15 occurrences of `variant("cxx", ...)`. What do you think about making it
```suggestion
    variant("c", default=False, description="Build C API")
```
?

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented on CMakeLists.txt:
```diff
@@ -44,6 +44,7 @@ option(DLAF_BUILD_TESTING "Build tests" ON)
 option(DLAF_BUILD_TESTING_HEADER "Build header tests" OFF)
 option(DLAF_BUILD_DOC "Build documentation" OFF)
 option(DLAF_WITH_PRECOMPILED_HEADERS "Use precompiled headers." OFF)
+option(DLAF_WITH_C_API "Build ScaLAPACK-like interface" OFF)
```
```suggestion
option(DLAF_WITH_C_API "Build C API (requires ScaLAPACK)" OFF)
```

RMeli commented on include/dlaf/communication/communicator_grid.h:
```diff
@@ -91,6 +91,11 @@ class CommunicatorGrid {
       return col_;
   }
 
+  /// Return full communicator order
+  dlaf::common::Ordering fullCommunicatorOrder() noexcept {
+    return FULL_COMMUNICATOR_ORDER;
+  }
+
```
```suggestion
```

RMeli commented on src/CMakeLists.txt:
```diff
@@ -274,6 +276,14 @@ DLAF_addSublibrary(
   LIBRARIES dlaf.core
 )
 
+# Define DLAF's interface library
```
```suggestion
# Define DLAF's C API library
```

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
I think this is now finally ready for another round of reviews. Thank you for the many and useful earlier comments/suggestions, which should have been addressed/implemented.

I suggest to address the following in follow-up PRs, to avoid further increasing the size and scope of this PR:
* Make ScaLAPACK an optional dependency (only for the ScaLAPACK-like C API)
* Add support for complex numbers
* Add generalized eigensolver

If there are no objections on the suggestion above, I'll convert this PR from draft to ready (as soon as CI comes back green).

RMeli commented:
cscs-ci run

RMeli commented on spack/packages/dla-future/package.py:
```diff
@@ -16,7 +16,10 @@ class DlaFuture(CMakePackage, CudaPackage, ROCmPackage):
     maintainers = ["rasolca", "albestro", "msimberg", "aurianer"]
 
-    version("0.1.0", sha256="f7ffcde22edabb3dc24a624e2888f98829ee526da384cd752b2b271c731ca9b1")
+    version(
+        "0.1.0",
+        sha256="f7ffcde22edabb3dc24a624e2888f98829ee526da384cd752b2b271c731ca9b1",
+    )
```
I applied `black` so some of the changes in `package.py` are un-related to the PR (and only consist in formatting).

RMeli replied:
Fixed formatting by using `black -l 99 -C`, according to Spack's `pyproject.toml` settings.

RMeli commented on spack/packages/dla-future/package.py:
```diff
@@ -25,11 +28,19 @@ class DlaFuture(CMakePackage, CudaPackage, ROCmPackage):
 
     variant("miniapps", default=False, description="Build miniapps.")
 
+    variant("c", default=False, description="Build C API")
```
Always compile the standard C API and make the ScaLAPACK API optional. Call this variant `+scalapack`.

RMeli commented:
cscs-ci run

RMeli commented on CMakeLists.txt:
```diff
@@ -44,6 +44,7 @@ option(DLAF_BUILD_TESTING "Build tests" ON)
 option(DLAF_BUILD_TESTING_HEADER "Build header tests" OFF)
 option(DLAF_BUILD_DOC "Build documentation" OFF)
 option(DLAF_WITH_PRECOMPILED_HEADERS "Use precompiled headers." OFF)
+option(DLAF_WITH_C_API "Build C API (requires ScaLAPACK)" OFF)
```
Changed to `DLAF_WITH_SCALAPACK`. The Spack variant is now `+scalapack` and the non-ScaLAPACK C API is always built.

RMeli commented:
black -l 99 -C

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

rasolca reviewed (commented):
- What about complex interface? will you add it in this PR?
- please use the `&&` and `||` operators (see https://github.com/eth-cscs/DLA-Future/blob/master/misc/coding.md#alternative-operator-representations)
- I'm always confused when I see `descx`. I usually think it is a scalapack descriptor but I realize later that is is a DLAF_descriptor. Any suggestion about improvint readability?
- For readability I would mark definitions of C function in cpp files with extern "C".

RMeli commented:
> What about complex interface? will you add it in this PR?

I was suggesting in a comment above to add the generalised eigensolver and the complex interface in a separate PR, to avoid bloating. If you prefer to have the complex interface here, I'm happy to add it.

> I'm always confused when I see descx. I usually think it is a scalapack descriptor but I realize later that is is a DLAF_descriptor. Any suggestion about improvint readability?

Yes, I agree it might be a bit confusing. What about `dlaf_descX`, for lack of better ideas?

RMeli commented:
> What about `dlaf_descX`, for lack of better ideas?

Actually, this is already the case for the eigensolver and the Cholesky decomposition. Was this confusing too or you meant in other parts of the code?

RMeli commented on src/c_api/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,122 @@
+    MatrixMirror eigenvalues(eigenvalues_host);
+
+    // TODO: Use dlaf_uplo instead of hard-coded blas::Uplo::Lower
+    // TODO: blas::Uplo::Uppper is not yet supported in DLA-Future
```
What do you mean here? It is a comment to remind to change it once `Upper` is supported. Or do you want to use `dlaf_uplo` directly here, and fail at the check in the main algorithm?

RMeli commented on spack/packages/dla-future/package.py:
```diff
@@ -116,16 +124,23 @@ def cmake_args(self):
+                self.define("MKL_LAPACK_TARGET", f"mkl::mkl_intel_32bit_{mkl_threads}_dyn"),
             ]
+            if "+scalapack" in spec:
+                if "^mpich" in spec or "^cray-mpich" in spec:
+                    mkl_mpi = "mpich"
+                elif "+openmpi" in spec:
+                    mkl_mpi = "omp"
```
Good point (it should also be `mkl_mpi = "ompi"`). The ScaLAPACK target in `FindMKL.cmake` is defined as
```
mkl::scalapack_[mpich|ompi]_[gf|intel]_[32bit|64bit]_[seq|omp|tbb]_[st|dyn]
```
so `mpich` and `ompi` are the only two options for the target. I'll check what it is suggested in the documentation for IntelMPI etc.

rasolca replied:
mpich and ompi are the two blacs variantsin mkl, however other mpi implementations are compatible with them:
e.g. intelmpi and mvapich are compatible with mpich.

RMeli replied:
I figured it was the case for IntelMPI and I added it already. I'll do the same for MVAPICH. I'm not sure if there is a more robust way of doing this.

RMeli replied:
The `intel-oneapi-mkl` package does something similar: https://github.com/spack/spack/blob/982cdd798858fb2048623c5c0833711a89bfbf51/var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py#L179-L182

RMeli commented:
cscs-ci run

RMeli commented on include/dlaf_c/factorization/cholesky.h:
```diff
@@ -0,0 +1,88 @@
+/// @param uplo Specify if upper ('U') or lower ('L') triangular part of @param a will be referenced
+/// @param n UNSUPPORTED, ignored
+/// @param a Local part of the global matrix
+/// @param ia UNSUPPORTED, ignored
```
Assertions added.

RMeli commented on include/dlaf_c/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,111 @@
+/// @param iz Global row index denoting the beginning of the submatrix to be operated on, UNSUPPORTED
+/// @param jz Global column index denoting the beginning of the submatrix to be operated on, UNSUPPORTED
+/// @param descz ScaLAPACK descriptor of @param z
+/// @param info 0 if the factorization completed normally
```
Thanks for the suggestion. I'm re-working the docs in order to improve consistency (also with ScaLAPACK).

RMeli replied:
I ended up going for `@copydoc`, which seems to do what we want (write doc once, copy paste it for all the functions with different names/argument types).

rasolca commented on src/c_api/factorization/cholesky.h:
```diff
@@ -0,0 +1,108 @@
+#include "../blacs.h"
+#include "../grid.h"
+
+void check_dlaf(char uplo, DLAF_descriptor desca);
```
not defined and not used anymore.

rasolca commented on include/dlaf_c/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,76 @@
+
+/// Eigensolver
+///
+/// The matrices $A$ and $Z$ are assumed to be distributed and in host memory.
```
Is $A$ displaying correctly in the doc?

RMeli replied:
I still need to build and double check the docs locally (but I think `\f` is missing).

rasolca commented:
> > What about `dlaf_descX`, for lack of better ideas?
> 
> Actually, this is already the case for the eigensolver and the Cholesky decomposition. Was this confusing too or you meant in other parts of the code?

The code I was referring to was removed in a later commit.

RMeli commented:
cscs-ci run

msimberg commented on include/dlaf_c/desc.h:
```diff
@@ -0,0 +1,24 @@
+
+#pragma once
+
+/// DLA-Future descriptor
```
```suggestion
/// DLA-Future matrix descriptor
```
?

msimberg commented on include/dlaf_c/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,76 @@
+DLAF_EXTERN_C void dlaf_pssyevd(char uplo, int n, float* a, int ia, int ja, int* desca, float* w,
+                                float* z, int iz, int jz, int* descz, int* info);
+
+/// \overload
```
Just a reminder to check this in the doxygen output. I _think_ this should be `Something something see @ref dlaf_pssyevd, something this is for doubles.` This is not strictly an overload of another function (since it can't be in C).

RMeli replied:
Good point. I looked into the Doxygen documentation (I haven't used it in a few yeas, hence I'm really rusty) and I think what we want is `@copydoc`.

msimberg commented on include/dlaf_c/factorization/cholesky.h:
```diff
@@ -0,0 +1,76 @@
+/// @param info 0 if the factorization completed normally
+DLAF_EXTERN_C void dlaf_pdpotrf(char uplo, int n, double* a, int ia, int ja, int* desca, int* info);
+
+/// \overload
```
Same here, reminder to check the documentation.

msimberg commented on include/dlaf_c/grid.h:
```diff
@@ -0,0 +1,67 @@
+
+#ifdef DLAF_WITH_SCALAPACK
+
+/// Create communication grid from BLACS context
+///
+/// Grids created here are indexed by their corresponding BLACS context
+/// @param blacs_ctxt.
+///
+/// The grid ordering is automatically inferred from the BLACS grid ordering.
+/// Only row-major and column-major grids are supported (created with
+/// blacs_gridinit). Grids created with blacs_gridmap are not supported.
+///
+/// @param blacs_ctxt BLACS context
```
I'm not sure we've had cases like this before, but do you think it would be useful to mention in the docstring that this function is only available if `DLAF_WITH_SCALAPACK=ON`? Also, docs should be built with `DLAF_WITH_SCALAPACK` defined for this to show up.

RMeli replied:
> Also, docs should be built with DLAF_WITH_SCALAPACK defined for this to show up.

Good point, this is done now with `PREDEFINED += DLAF_WITH_SCALAPACK`.

RMeli replied:
> but do you think it would be useful to mention in the docstring that this function is only available if DLAF_WITH_SCALAPACK=ON

Yes, I think it would be useful! I added this under a `@remark`.

msimberg commented on include/dlaf_c/init.h:
```diff
@@ -0,0 +1,27 @@
+
+/// Initialize pika runtime and DLA-Future
+///
+/// The pika runtime is automatically suspended within this function
```
Make post condition (with `@post`)?

RMeli replied:
Started going through the documentation and added special keywords where sensible (including this one).

msimberg commented on include/dlaf_c/utils.h:
```diff
@@ -0,0 +1,27 @@
+#define DLAF_EXTERN_C
+#endif
+
+#ifdef __cplusplus
+#include <complex>
+using complex_c = std::complex<float>;
+using complex_z = std::complex<double>;
+#else
+#include <complex.h>
+typedef float complex complex_c;
+typedef double complex complex_z;
+#endif
```
```suggestion
#ifdef __cplusplus
#include <complex>
using dlaf_complex_c = std::complex<float>;
using dlaf_complex_z = std::complex<double>;
#else
#include <complex.h>
typedef float complex dlaf_complex_c;
typedef double complex dlaf_complex_z;
#endif
```
to avoid potential conflicts since the names are otherwise quite generic? I guess these should not be visible to users anyway, but these may be included inside DLAF with something that uses those names.

msimberg commented on test/unit/c_api/factorization/test_cholesky_c_api.cpp:
```diff
@@ -0,0 +1,244 @@
+const std::vector<blas::Uplo> blas_uplos({blas::Uplo::Lower, blas::Uplo::Upper});
+
+const std::vector<std::tuple<SizeType, SizeType>> sizes = {
+    //{0, 2},                              // m = 0
+    //{5, 8}, {34, 34},                    // m <= mb
+    //{4, 3}, {16, 10}, {34, 13}, {32, 5}  // m > mb
```
To do: uncomment?

RMeli replied:
Thanks for spotting this, I forgot. These currently do not work because result in some ranks without local tiles and therefore `LocalTileIndex(0,0)` fails (I use this to extract the memory of DLAF-allocated matrices). I'll need to fix this.

msimberg replied:
Closing this one since there's another comment covering the same: https://github.com/eth-cscs/DLA-Future/pull/886#discussion_r1258231549.

msimberg commented on test/unit/c_api/eigensolver/test_eigensolver_c_api.cpp:
```diff
@@ -0,0 +1,256 @@
+  for (const comm::CommunicatorGrid& grid : this->commGrids()) {
+    for (auto uplo : blas_uplos) {
+      for (auto [m, mb, b_min] : sizes) {
+        // getTuneParameters().eigensolver_min_band = b_min;
```
To do: uncomment?

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented on spack/packages/dla-future/package.py:
```diff
@@ -116,16 +119,29 @@ def cmake_args(self):
+                    or "^mvapich2" in spec
+                ):
+                    mkl_mpi = "mpich"
+                elif "+openmpi" in spec:
```
```suggestion
                elif "^openmpi" in spec:
```

RMeli commented:
cscs-ci run

albestro commented on test/include/dlaf_test/config_c_api.h.in:
```diff
@@ -0,0 +1,28 @@
+// Configuration for C API tests
+// PIKA parameters from CMake to be used in C API tests
+// C API tests are not registered as PIKA tests since the PIKA runtime is
+// initialized and stipped for every test
```
?
```suggestion
// initialized and stopped for every test
```

albestro commented on test/unit/c_api/factorization/test_cholesky_c_api.cpp:
```diff
@@ -0,0 +1,243 @@
+const std::vector<blas::Uplo> blas_uplos({blas::Uplo::Lower, blas::Uplo::Upper});
+
+const std::vector<std::tuple<SizeType, SizeType>> sizes = {
+    //    {0, 2},  // m = 0
+    // {5, 8}, {34, 34},                    // m <= mb
+    // {4, 3}, {16, 10}, {34, 13}, {32, 5},  // m > mb
+    {4, 1},
+    //   {34, 13},
+    {32, 5}  // m > mb
```
Why are these test-cases disabled?

RMeli replied:
This is what I'm working on at the moment, to re-enable these tests too. The reason for now is that for such cases `mat_h.readwrite(LocalTileIndex(0, 0))` fails.

albestro commented on cmake/template/DLAFConfig.cmake.in:
```diff
@@ -61,10 +62,20 @@ if(DLAF_WITH_MKL)
   find_dependency(LAPACK)
+
+  if(DLAF_WITH_SCALAPACK)
+    find_dependency(SCALAPACK)
+    set(SCALAPACK_LIBRARY "@SCALAPACK_LIBRARY@")
```
shouldn't these two lines be inverted?

albestro commented on cmake/FindSCALAPACK.cmake:
```diff
@@ -0,0 +1,39 @@
+  PATH_SUFFIXES lib
+)
+
+find_package_handle_standard_args(SCALAPACK DEFAULT_MSG SCALAPACK_LIBRARIES)
+
+mark_as_advanced(SCALAPACK_LIBRARY)
+
+if(SCALAPACK_FOUND)
+  if(NOT TARGET DLAF::SCALAPACK)
+    add_library(DLAF::SCALAPACK INTERFACE IMPORTED GLOBAL)
+  endif()
+
+  target_link_libraries(DLAF::SCALAPACK INTERFACE "${SCALAPACK_LIBRARY}")
+endif()
```
Here there is a mixup of `_LIBRARY` and `_LIBRARIES`. I commented also in the other file, but actually it's up to you deciding the final name for the variable 😄

albestro commented on include/dlaf_c/desc.h:
```diff
@@ -0,0 +1,26 @@
+  int jsrc;  ///< Process column of the first column of the global matrix
+  int i;     ///< First row of the submatrix within global matrix, has to be 1
+  int j;     ///< First column of the submatrix within global matrix, has to be 1
+  int ld;    ///< Leading fimension of the local matrix
```
```suggestion
  int ld;    ///< Leading dimension of the local matrix
```

albestro commented on include/dlaf_c/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,102 @@
+/// @param a Local part of the global matrix \f$\mathbf{A}\f$
+/// @param ia row index of the global matrix \f$\mathbf{A}\f$ identifying the first row of the submatrix
+/// $A$, has to be 1
+/// @param ja column index of the global matrix \f$\mathbf{A}\f$ identifying the firs column of the
```
```suggestion
/// @param ja column index of the global matrix \f$\mathbf{A}\f$ identifying the first column of the
```

albestro commented on include/dlaf_c/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,102 @@
+/// @param z Local part of the global matrix \f$\mathbf{Z}\f$
+/// @param iz row index of the global matrix \f$\mathbf{Z}\f$ identifying the first row of the submatrix
+/// \f$\mathbf{Z}\f$, has to be 1
+/// @param jz column index of the global matrix \f$\mathbf{A}\f$ identifying the firs column of the
```
```suggestion
/// @param jz column index of the global matrix \f$\mathbf{A}\f$ identifying the first column of the
```

albestro commented on include/dlaf_c/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,102 @@
+/// @param jz column index of the global matrix \f$\mathbf{A}\f$ identifying the firs column of the
+/// submatrix \f$\mathbf{A}\f$, has to be 1
+/// @param descz ScaLAPACK array descriptor of the global matrix \f$\mathbf{Z}\f$
+/// @param info 0 if the eigensolver completed normally
```
probably is clear for who is used to this, but... might it be worth specifying it is a return value?

RMeli replied:
`@return` does not take a name, I think the best option is `@param[out]`?

albestro replied:
I'm not super proficient with oxygen, but it looks like the right choice 😉

albestro commented on include/dlaf_c/factorization/cholesky.h:
```diff
@@ -0,0 +1,83 @@
+/// @param a Local part of the global matrix \f$\mathbf{A}\f$
+/// @param ia row index of the global matrix \f$\mathbf{A}\f$ identifying the first row of the submatrix
+/// \f$\mathbf{A}\f$, has to be 1
+/// @param ja column index of the global matrix \f$\mathbf{A}\f$ identifying the firs column of the
```
```suggestion
/// @param ja column index of the global matrix \f$\mathbf{A}\f$ identifying the first column of the
```

albestro commented on include/dlaf_c/factorization/cholesky.h:
```diff
@@ -0,0 +1,83 @@
+/// @param ja column index of the global matrix \f$\mathbf{A}\f$ identifying the firs column of the
+/// submatrix \f$\mathbf{A}\f$, has to be 1
+/// @param desca ScaLAPACK array descriptor of the global matrix \f$\mathbf{A}\f$
+/// @param info 0 if the factorization completed normally
```
just for reference, here too the "document it is a return type".

albestro commented on include/dlaf_c/grid.h:
```diff
@@ -0,0 +1,69 @@
+/// @warning Grids created here are indexed starting from INT_MAX, to avoid clashes with
+/// BLACS contexts (which start from 0).
+///
+/// Grid ordering can be column-major ("C") or rowp-major ("R").
```
```suggestion
/// Grid ordering can be column-major ("C") or row-major ("R").
```

albestro commented on include/dlaf_c/init.h:
```diff
@@ -0,0 +1,29 @@
+DLAF_EXTERN_C void dlaf_initialize(int argc_pika, const char** argv_pika, int argc_dlaf,
+                                   const char** argv_dlaf);
+
+/// Finalize DLA-Future and pika runtime
```
and when `dlaf_finalize` returns the pika runtime is resumed, right?

In any case, I would specify the behaviour in the doc (as you already did with `@post` for `dlaf_initialize`.

RMeli replied:
The `pika` runtime is resumed when `dlaf_finalize` is called, then it is finalized and stopped (not suspended). I'll improve the documentation.

albestro commented on include/dlaf_c/init.h:
```diff
@@ -0,0 +1,29 @@
+
+#include <dlaf_c/utils.h>
+
+/// Initialize pika runtime and DLA-Future
+///
+/// @post The pika runtime is automatically suspended when this function returns
```
Would it be superfluous saying "if not already initialised"?

albestro commented on src/c_api/init.cpp:
```diff
@@ -0,0 +1,47 @@
+  dlaf::finalize();
+  pika::stop();
+
+  dlaf_initialized = false;
```
no check needed about dlaf/pika being initialised?

albestro commented on test/unit/c_api/eigensolver/test_eigensolver_c_api.cpp:
```diff
@@ -0,0 +1,276 @@
+    pika::suspend();
+
+    if constexpr (api == API::dlaf) {
+      DLAF_descriptor dlaf_desc_a = {(int) m, (int) m, (int) mb, (int) mb, 0, 0, 1, 1, lld_a};
+      DLAF_descriptor dlaf_desc_eigenvectors = {(int) m, (int) m, (int) mb, (int) mb,        0,
+                                                0,       1,       1,        lld_eigenvectors};
```
minor: not sure, but you might use `()` instead of `{}` in order to allow narrowing conversions.

RMeli replied:
I tend to use `{}` to avoid narrowing conversion, but I agree in this case doing them explicitly does not make much sense...

albestro commented on test/unit/c_api/eigensolver/test_eigensolver_c_api.cpp:
```diff
@@ -0,0 +1,276 @@
+  if (!mat_a_h.size().isEmpty())
+    testEigensolverCorrectness(uplo, reference, ret.eigenvalues, ret.eigenvectors, grid);
+
+  // Suspend pika to make sure dla_finalize resumes it
```
```suggestion
  // Suspend pika to make sure dlaf_finalize resumes it
```

albestro commented on test/unit/c_api/factorization/test_cholesky_c_api.cpp:
```diff
@@ -0,0 +1,243 @@
+
+    local_a_ptr = toplefttile_a.ptr();
+    lld = static_cast<int>(toplefttile_a.ld());
+  }  // Destroy tile (avoids deoendency issues down the line)
```
```suggestion
  }  // Destroy tile (avoids dependency issues down the line)
```

albestro commented on test/unit/c_api/factorization/test_cholesky_c_api.cpp:
```diff
@@ -0,0 +1,243 @@
+  CHECK_MATRIX_NEAR(res, mat_h, 4 * (mat_h.size().rows() + 1) * TypeUtilities<T>::error,
+                    4 * (mat_h.size().rows() + 1) * TypeUtilities<T>::error);
+
+  // Suspend pika to make sure dla_finalize resumes it
```
```suggestion
  // Suspend pika to make sure dlaf_finalize resumes it
```

albestro commented on src/c_api/grid.cpp:
```diff
@@ -0,0 +1,93 @@
+std::unordered_map<int, dlaf::comm::CommunicatorGrid> dlaf_grids;
+
+int dlaf_create_grid(MPI_Comm comm, int nprow, int npcol, char order) {
+  // dlaf_context starts from INT_MAX to reeduce the likelihood of clashes with blacs contexts
```
```suggestion
  // dlaf_context starts from INT_MAX to reduce the likelihood of clashes with blacs contexts
```

albestro commented on src/c_api/grid.cpp:
```diff
@@ -0,0 +1,93 @@
+  MPI_Allreduce(&_col_major, &col_major, 1, MPI_C_BOOL, MPI_LAND, comm);
+
+  if (!row_major and !col_major) {
+    std::cout << "Grid layout must be row major or columns majord." << std::endl;
```
```suggestion
    std::cout << "Grid layout must be row major or column major." << std::endl;
```

albestro commented on spack/packages/dla-future/package.py:
```diff
@@ -138,6 +154,11 @@ def cmake_args(self):
                 )
             )
+            if "+scalapack" in spec:
+                args.append(self.define("SCALAPACK_LIBRARY", spec["scalapack"].libs.ld_flags))
```
Be careful here too with the name of the variable.

RMeli replied:
This should be consistent with the `LAPACK_LIBRARY` variable above.

albestro commented on include/dlaf_c/desc.h:
```diff
@@ -0,0 +1,26 @@
+  int nb;    ///< Column blocking factor
+  int isrc;  ///< Process row of the first row of the global matrix
+  int jsrc;  ///< Process column of the first column of the global matrix
+  int i;     ///< First row of the submatrix within global matrix, has to be 1
+  int j;     ///< First column of the submatrix within global matrix, has to be 1
```
Last comment: it is not clear to me why 1.

This is not SCALAPACK dependent but C-API, so I would say we might want it 0-based. Is there any reason why is it better to have them 1-based?

RMeli commented on src/c_api/utils.cpp:
```diff
@@ -0,0 +1,21 @@
+
+#include <dlaf/common/assert.h>
+
+blas::Uplo dlaf_uplo_from_char(char uplo) {
```
Ignorance ;)

RMeli commented:
cscs-ci run

msimberg commented on include/dlaf_c/factorization/cholesky.h:
```diff
@@ -0,0 +1,83 @@
+/// @param ja column index of the global matrix \f$\mathbf{A}\f$ identifying the first column of the
+/// submatrix \f$\mathbf{A}\f$, has to be 1
+/// @param desca ScaLAPACK array descriptor of the global matrix \f$\mathbf{A}\f$
+/// @param info 0 if the factorization completed normally
```
I'm not sure we can do anything about this already now, but I guess `info` will always be `0` or the algorithm will terminate? We don't currently have a way to handle all errors in algorithms and report them to callers (only those happening at schedule time).

One potential direction for this would be to wrap algorithm execution in something like an `async_scope` and then explicitly wait for all work in the `async_scope` (and check errors) before returning from the C API. This probably deserves its own issue so that we don't forget about it.

No action required for this PR.

msimberg commented on src/c_api/factorization/cholesky.h:
```diff
@@ -0,0 +1,118 @@
+
+    return 0;
+  }
+  catch (std::exception& e) {
+    std::cerr << e.what() << '\n';
+    return -1;
+  }
```
I think most exceptions will inherit from `std::exception` but they're not guaranteed to do so. We could either:
- do what you do now, and non-`std::exception`s will terminate at the end of the function because it's `noexcept`, or
- add another catch-all catch:
```suggestion
  catch (const std::exception& e) {
    std::cerr << e.what() << '\n';
    return -1;
  }
  catch (...)
  {
    std::cerr << "blabla unknown exception thrown in dlaf_xxx\n";
    return -1;
  }
```
I think both are fine, but I'm slightly in favour of catching all exceptions and let the caller deal with the `-1`.

RMeli commented:
cscs-ci run

albestro commented on include/dlaf_c/init.h:
```diff
@@ -26,4 +28,8 @@ DLAF_EXTERN_C void dlaf_initialize(int argc_pika, const char** argv_pika, int ar
 
 /// Finalize DLA-Future and pika runtime
+///
+/// @post The pika runtime is fnialized and stopped when this function returns
```
```suggestion
/// @post The pika runtime is finalized and stopped when this function returns
```

RMeli replied:
Thanks! I need to setup a spell-checker...

albestro commented on src/c_api/eigensolver/eigensolver.h:
```diff
@@ -24,56 +25,64 @@
+    pika::suspend();
+    return 0;
+  }
+  catch (std::exception& e) {
```
```suggestion
  catch (const std::exception& e) {
```

albestro commented on src/c_api/factorization/cholesky.h:
```diff
@@ -22,83 +24,98 @@
-  pika::suspend();
+    return 0;
+  }
+  catch (std::exception& e) {
```
```suggestion
  catch (const std::exception& e) {
```

albestro commented on test/include/dlaf_c_test/config.h.in:
```diff
@@ -0,0 +1,28 @@
+// Configuration for C API tests
+// PIKA parameters from CMake to be used in C API tests
+// C API tests are not registered as PIKA tests since the PIKA runtime is
+// initialized and stipped for every test
```
```suggestion
// initialized and stopped for every test
```

RMeli commented on test/include/dlaf_test/comm_grids/grids_6_ranks.h:
```diff
@@ -67,6 +67,25 @@ class CommunicatorGrid6RanksEnvironment : public ::testing::Environment {
+      if (world.size() == 1)
+        return;
+
+      comm_grids.emplace_back(world, 3, 2, common::Ordering::RowMajor);
+      comm_grids.emplace_back(world, 2, 3, common::Ordering::RowMajor);
+      comm_grids.emplace_back(world, 2, 3, common::Ordering::ColumnMajor);
+      comm_grids.emplace_back(world, 3, 2, common::Ordering::ColumnMajor);
```
Not necessarily, but the main issue I encountered here is that the communicator stored within the `grid` is always row-major. Therefore, to get the grid ordering I needed to used the original communicator `MPI_COMM_WORLD`.

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

rasolca commented on src/c_api/eigensolver/eigensolver.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void dlaf_pssyevd(char uplo, int m, float* a, int ia, int ja, int* desca, float* w, float* z, int iz,
+                  int jz, int* descz, int* info) {
+  pxxxevd<float>(uplo, m, a, ia, ja, desca, w, z, iz, jz, descz, *info);
```
Hermitian is equivalent to symmetric for real cases.
```suggestion
  pxheevd<float>(uplo, m, a, ia, ja, desca, w, z, iz, jz, descz, *info);
```

RMeli replied:
Yes, it was to be consistent with ScaLAPACK naming convention. I can change it preferred.

rasolca commented on src/c_api/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,154 @@
+    dlaf::TileElementSize block_size(dlaf_desca.mb, dlaf_desca.nb);
+
+    dlaf::comm::Index2D src_rank_index(dlaf_desca.isrc, dlaf_desca.jsrc);
+
+    dlaf::matrix::Distribution distribution(matrix_size, block_size, communicator_grid.size(),
+                                            communicator_grid.rank(), src_rank_index);
+
+    dlaf::matrix::LayoutInfo layout_a = colMajorLayout(distribution, dlaf_desca.ld);
+    dlaf::matrix::LayoutInfo layout_z = colMajorLayout(distribution, dlaf_descz.ld);
+
+    MatrixHost matrix_host(distribution, layout_a, a);
+    MatrixHost eigenvectors_host(distribution, layout_z, z);
+    auto eigenvalues_host = dlaf::matrix::createMatrixFromColMajor<dlaf::Device::CPU>(
+        {dlaf_descz.m, 1}, {distribution.blockSize().rows(), 1}, std::max(dlaf_descz.m, 1), w);
+
+    {
+      MatrixMirror matrix(matrix_host);
+      MatrixMirror eigenvectors(eigenvectors_host);
+      MatrixBaseMirror eigenvalues(eigenvalues_host);
+
+      dlaf::eigensolver::eigensolver<dlaf::Backend::Default, dlaf::Device::Default, T>(
+          communicator_grid, blas::char2uplo(uplo), matrix.get(), eigenvalues.get(), eigenvectors.get());
+    }  // Destroy mirror
+
+    // Ensure data is copied back to the host
+    eigenvalues_host.waitLocalTiles();
+    eigenvectors_host.waitLocalTiles();
+
+    pika::suspend();
+    return 0;
+  }
+  catch (const std::exception& e) {
+    std::cerr << e.what() << '\n';
+    return -1;
+  }
+  catch (...) {
+    std::cerr << "ERROR: Unknown exception caught in DLA-Future's eigensolver." << '\n';
+    return -1;
+  }
+}
```
I don't get what it is supposed to throw an exception that has to handled.
All the DLAF checks terminate if not fulfilled.

rasolca commented on src/c_api/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,154 @@
+  }
+  catch (const std::exception& e) {
+    std::cerr << e.what() << '\n';
+    info = -1;
```
As above about exceptions.
As `info = -1` in lapack and scalapack means that the first function argument is wrong, is this the exception you want to catch?

rasolca commented on test/unit/c_api/eigensolver/test_eigensolver_c_api_wrapper.c:
```diff
@@ -0,0 +1,57 @@
+#include <dlaf_c/utils.h>
+
+#ifdef DLAF_WITH_SCALAPACK
+void C_dlaf_pdsyevd(char uplo, int m, double* a, int ia, int ja, int* desca, double* w, double* z,
+                    int iz, int jz, int* descz, int* info) {
+  dlaf_pdsyevd(uplo, m, a, ia, ja, desca, w, z, iz, jz, descz, info);
+}
+
+void C_dlaf_pssyevd(char uplo, int m, float* a, int ia, int ja, int* desca, float* w, float* z, int iz,
+                    int jz, int* descz, int* info) {
+  dlaf_pssyevd(uplo, m, a, ia, ja, desca, w, z, iz, jz, descz, info);
+}
+
+void C_dlaf_pzheevd(char uplo, int m, dlaf_complex_z* a, int ia, int ja, int* desca, double* w,
+                    dlaf_complex_z* z, int iz, int jz, int* descz, int* info) {
+  dlaf_pzheevd(uplo, m, a, ia, ja, desca, w, z, iz, jz, descz, info);
+}
+
+void C_dlaf_pcheevd(char uplo, int m, dlaf_complex_c* a, int ia, int ja, int* desca, float* w,
+                    dlaf_complex_c* z, int iz, int jz, int* descz, int* info) {
+  dlaf_pcheevd(uplo, m, a, ia, ja, desca, w, z, iz, jz, descz, info);
+}
```
What is the reason for this wrappers?

RMeli replied:
See discussion in https://github.com/eth-cscs/DLA-Future/pull/886#discussion_r1211306641. Do I need to add a comment?

rasolca replied:
Yes please.

rasolca commented on src/c_api/factorization/cholesky.h:
```diff
@@ -0,0 +1,121 @@
+             int& info) noexcept {
+  try {
+    using MatrixMirror = dlaf::matrix::MatrixMirror<T, dlaf::Device::Default, dlaf::Device::CPU>;
+
+    DLAF_ASSERT(desca[0] == 1, desca[0]);
+    DLAF_ASSERT(ia == 1, ia);
+    DLAF_ASSERT(ja == 1, ja);
+
+    pika::resume();
+
+    // Get grid corresponding to blacs context in desca
+    // The grid needs to be created with dlaf_create_grid_from_blacs
+    auto communicator_grid = dlaf_grids.at(desca[1]);
+    dlaf::matrix::Distribution distribution({n, n}, {desca[4], desca[5]}, communicator_grid.size(),
+                                            communicator_grid.rank(), {desca[6], desca[7]});
+    dlaf::matrix::LayoutInfo layout_info = colMajorLayout(distribution, desca[8]);
+
+    dlaf::matrix::Matrix<T, dlaf::Device::CPU> matrix_host(std::move(distribution), layout_info, a);
+
+    {
+      MatrixMirror matrix(matrix_host);
+
+      dlaf::factorization::cholesky<dlaf::Backend::Default, dlaf::Device::Default, T>(
+          communicator_grid, blas::char2uplo(uplo), matrix.get());
+    }  // Destroy mirror
+
+    matrix_host.waitLocalTiles();
+
+    pika::suspend();
+
+    info = 0;
+  }
+  catch (const std::exception& e) {
+    std::cerr << e.what() << '\n';
+    info = -1;
+  }
+  catch (...) {
+    std::cerr << "ERROR: Unknown exception caught in DLA-Future's Cholesky decomposition." << '\n';
+    info = -1;
+  }
```
Here there is a lot of code duplication (w.r.t the above `cholesky` function).
I would create a function: `make_dlaf_descriptor(m, n, i, j, desc)` to extract the needed information from the Scalapack descriptor and then call the above `cholesky` function.

rasolca commented on src/c_api/factorization/cholesky.h:
```diff
@@ -0,0 +1,121 @@
+
+    auto communicator_grid = dlaf_grids.at(dlaf_context);
+
+    dlaf::GlobalElementSize matrix_size(dlaf_desca.m, dlaf_desca.n);
+    dlaf::TileElementSize block_size(dlaf_desca.mb, dlaf_desca.nb);
+
+    dlaf::comm::Index2D src_rank_index(dlaf_desca.isrc, dlaf_desca.jsrc);
+
+    dlaf::matrix::Distribution distribution(matrix_size, block_size, communicator_grid.size(),
+                                            communicator_grid.rank(), src_rank_index);
+
+    dlaf::matrix::LayoutInfo layout = colMajorLayout(distribution, dlaf_desca.ld);
+
+    dlaf::matrix::Matrix<T, dlaf::Device::CPU> matrix_host(std::move(distribution), layout, a);
```
I would consider to include this code in a function such that it can be reused in all the c wrappers.

RMeli commented:
cscs-ci run

RMeli commented on include/dlaf_c/utils.h:
```diff
@@ -0,0 +1,33 @@
+
+#include <dlaf_c/desc.h>
+
+DLAF_EXTERN_C struct DLAF_descriptor make_dlaf_descriptor(int m, int n, int i, int j, int* desc);
```
Add documentation.

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

RMeli commented on src/c_api/eigensolver/eigensolver.h:
```diff
@@ -0,0 +1,93 @@
+  auto communicator_grid = dlaf_grids.at(dlaf_context);
+
+  auto [distribution_a, layout_a] = distribution_and_layout(dlaf_desca, communicator_grid);
+  auto [distribution_z, layout_z] = distribution_and_layout(dlaf_desca, communicator_grid);
```
```suggestion
  auto [distribution_z, layout_z] = distribution_and_layout(dlaf_descz, communicator_grid);
```

RMeli commented on test/unit/c_api/eigensolver/test_eigensolver_c_api.cpp:
```diff
@@ -0,0 +1,336 @@
+  pika::suspend();
+}
+
+TYPED_TEST(EigensolverTestMC, pika) {
+  for (int i = 0; i < 1000; i++) {
+    pika::start(nullptr, 0, nullptr);
+    pika::finalize();
+    pika::stop();
+  }
+}
```
Remove when done.

RMeli commented:
```
TYPED_TEST(EigensolverTestMC, pika) {
  for (int i = 0; i < 1000; i++) {
    pika::start(nullptr, 0, nullptr);
    pika::finalize();
    pika::stop();
  }
}
```
[fails with `pika@main`](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/4675576268#L2317) but [works with `pika@git.fix-free_stack-no-guard-pages`](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/pipelines/935042591), so https://github.com/pika-org/pika/pull/716 appears to solve the issue. Thanks @msimberg!

RMeli commented:
cscs-ci run

RMeli commented:
cscs-ci run

mtaillefumier commented on cmake/FindSCALAPACK.cmake:
we need to namespace the custom cmake modules with dlaf_ so that we do not enter in conflict with other findscalapack.cmake

mtaillefumier reviewed (approved):
I think the code was scrutinized so many times over by the group that I skipped most of the cpp and c code.

RMeli commented:
- [ ] Update according to https://github.com/eth-cscs/DLA-Future/pull/942

msimberg commented on include/dlaf_c/grid.h:
```diff
@@ -0,0 +1,70 @@
+#ifdef DLAF_WITH_SCALAPACK
+
+/// Create communication grid from BLACS context
+///
```
```suggestion
///
/// @remark This function is available only when DLAF_WITH_SCALAPACK=ON.
///
```
? Probably worth double checking other `#ifdef DLAF_WITH_SCALAPACK` docstrings as well.

RMeli replied:
Good catch, thanks!

msimberg reviewed (approved):
I have no further comments on this. Thanks @RMeli for all the work on this!

rasolca commented:
cscs-ci run

rasolca reviewed (approved):
Looks good, thanks!

RMeli commented:
Thanks you all for the extensive reviews and for all the suggestions!
